src/backend/api/meals.js

```diff
@@ -1,14 +1,288 @@
-const express = require("express");
+const express = require('express');
+const { sum, limit } = require('../database');
 const router = express.Router();
-const knex = require("../database");
+const knex = require('../database');
 
-router.get("/", async (request, response) => {
+// to validate data received from user to post/Put to db
+
+const mealColumns = new Set([
+  'title',
+  'description',
+  'location',
+  'when',
+  'max_reservations',
+  'price',
+  'created_date',
+]);
+
+function checkValidData(data) {
+  try {
+    let validData = {};
+    for (const key in data) {
+      // checking if the mealColumns has the key
+      if (mealColumns.has(key)) {
+        validData[key] = data[key];
+      }
+    }
+    if ('limit' in validData) {
+      if (typeof validData.limit !== 'number') {
+        throw new Error();
+      }
+    }
+    if ('price' in validData) {
+      if (typeof validData.price !== 'number') {
+        throw new Error();
+      }
+    }
+    if ('max_reservation' in validData) {
+      if (typeof validData.max_reservation !== 'number') {
+        throw new Error();
+      }
+    }
+    if ('when' in validData) {
+      const parsedDate = new Date(validData.when);
+      if (isNaN(parsedDate)) {
+        throw new Error();
+      }
+    }
+    return validData;
+  } catch (error) {
+    console.log(error);
+    response.status(500).json({
+      status: 'failed',
+      message: `internal server error in checkValidData ${error}`,
+    });
+  }
+}
+
+async function getMeals(query) {
```

Comment:
While it works as expected there is one case that could be done better and if you're up for it you can improve it:

- The route will not work with multiple parameters at the same time as expected. If for instance I pass in both 'maxPrice' and 'availableReservations' you will return meals filtered only by the MaxPrice parameters since that is the one you check first.
- The first thing you are doing is fetching ALL of the meals from the db by: ```  let mealsArray = await knex('meal');``` and then filter them. This is not the most efficient way to do it. In your case it works perfectly fine, but if you had millions of records in the database, it could turn out to be slow!

I strongly suggest you watch the following [video ](https://www.youtube.com/watch?v=4KV79W4g2QU&t=418s)made by Orhan at HYF which explains how you can achieve the optimal behavior with the multiple params but also having an optimized route which only fetches what is necessary from the database.

Reply:
hi @dsamourkasidis  , i have fix those issues you have pointed? is this ok? please  have a look one time.

Reply:
Looked great good job 👌